File: js/sorting_algos/heap_sort.js
// building a sample array of size (n)
var a = []
var nSample = Math.round(Math.random() * 100)
for (var i=0; i < nSample; i++) {
  a.push(Math.round(Math.random() * nSample))
}

const switchEm = (arr, i, j) => {
  var curr = arr[i];
  arr[i] = arr[j];
  arr[j] = curr;
}

const sortDown = (arr, i, max) => {
  var i_large, a1, a2;
  while(i < max) {
    i_large = i;
    a1 = 2 * i + 1;
    a2 = a1 + 1;
    if (a1 < max && arr[a1] > arr[i_large]) {
      i_large = a1;
    }
    if (a2 < max && arr[a2] > arr[i_large]) {
      i_large = a2;
    }
    if (i_large == i) {return};
    switchEm(arr, i, i_large);
    i = i_large;
  }
}

const heapOrder = (arr) => {
  var i = arr.length / 2 - 1;
  i = Math.floor(i);
  while (i >= 0) {
    sortDown(arr, i, arr.length);
    i -= 1;
  }
}

const heapSort = (arr) => {
  heapOrder(arr);
  finish = arr.length - 1;
  while(finish > 0) {
    switchEm(arr, 0, finish);
    sortDown(arr, 0, finish);
    finish -= 1
  }
  return arr;
}

File: py/sorting_algos/heap_sort.py
import random
n = random.randint(1, 1000)
ab = []
for i in range(n):
  ab.append(random.randint(1, n))

def sort_down(a, start, finish):
  base = start
  while True:
    child = base * 2+1
    if child > finish: break
    if child + 1 <= finish and a[child] < a[child+1]:
      child += 1
    if a[base] < a[child]:
      a[base], a[child] = a[child], a[base]
      base = child
    else:
      break

def heap_sort(a):
  for start in range((len(a)-2/2), -1, -1):
    sort_down(a, start, len(a)-1)
  for finish in range(len(a)-1, 0, -1):
    a[finish], a[0] = a[0], a[finish]
    sort_down(a, 0, finish - 1)
  return a

print heap_sort(ab)